File: packages/typescript/service/src/disclosure.ts
/**
 * Caller-class-aware disclosure resolution.
 *
 * Two modes:
 * - Fixed mode (disclosureLevel != "policy"): returns the fixed level.
 * - Policy mode ("policy"): resolves from token claims via disclosurePolicy.
 */

const SCOPE_TO_CLASS: Record<string, string> = {
  "audit:full": "audit_full",
};

function resolveCallerClass(
  tokenClaims: Record<string, unknown> | null,
): string {
  if (tokenClaims == null) return "default";

  const callerClass = tokenClaims["anip:caller_class"];
  if (callerClass != null) return String(callerClass);

  const scopes = tokenClaims.scope;
  if (Array.isArray(scopes)) {
    for (const scope of scopes) {
      if (typeof scope === "string" && scope in SCOPE_TO_CLASS) {
        return SCOPE_TO_CLASS[scope];
      }
    }
  }

  return "default";
}

export function resolveDisclosureLevel(
  disclosureLevel: string,
  tokenClaims: Record<string, unknown> | null,
  disclosurePolicy?: Record<string, string>,
): string {
  if (disclosureLevel !== "policy") return disclosureLevel;

  const callerClass = resolveCallerClass(tokenClaims);

  if (disclosurePolicy == null) return "redacted";

  const level = disclosurePolicy[callerClass];
  if (level != null) return level;

  return disclosurePolicy["default"] ?? "redacted";
}

File: packages/go/service/disclosure.go
package service

// ResolveDisclosureLevel resolves the effective disclosure level.
// Fixed modes ("full", "reduced", "redacted") pass through.
// "policy" mode resolves caller class from token claims and applies the policy map.
// Implements SPEC §6.9 caller-class-aware disclosure.
func ResolveDisclosureLevel(level string, tokenClaims map[string]any, policy map[string]string) string {
	if level != "policy" {
		return level
	}

	callerClass := resolveCallerClass(tokenClaims)

	if policy == nil {
		return "redacted"
	}

	if mapped, ok := policy[callerClass]; ok {
		return mapped
	}
	if def, ok := policy["default"]; ok {
		return def
	}
	return "redacted"
}

// resolveCallerClass determines the caller class from token claims.
// Priority: 1) explicit anip:caller_class claim, 2) scope-derived, 3) "default".
func resolveCallerClass(claims map[string]any) string {
	if claims == nil {
		return "default"
	}

	if cc, ok := claims["anip:caller_class"].(string); ok && cc != "" {
		return cc
	}

	if scopes, ok := claims["scope"].([]any); ok {
		for _, s := range scopes {
			if str, ok := s.(string); ok && str == "audit:full" {
				return "audit_full"
			}
		}
	}

	return "default"
}
